File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/model/GitConfiguration.java
package net.nemerosa.ontrack.extension.git.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import net.nemerosa.ontrack.extension.issues.model.ConfiguredIssueService;
import net.nemerosa.ontrack.git.GitRepository;
import net.nemerosa.ontrack.model.support.UserPassword;

import java.util.Optional;

/**
 * Definition of a Git configuration.
 */
public interface GitConfiguration {

    /**
     * Type
     */
    String getType();

    /**
     * Name in the type
     */
    String getName();

    /**
     * Remote URL-ish
     */
    String getRemote();

    /**
     * Credentials
     */
    Optional<UserPassword> getCredentials();

    /**
     * Link to a commit, using {commit} as placeholder
     */
    String getCommitLink();

    /**
     * Link to a file at a given commit, using {commit} and {path} as placeholders
     */
    String getFileAtCommitLink();

    /**
     * Indexation interval
     */
    int getIndexationInterval();

    /**
     * Gets the associated issue service configuration (if any)
     */
    Optional<ConfiguredIssueService> getConfiguredIssueService();

    /**
     * Gets the Git repository for this configuration
     */
    @JsonIgnore
    default GitRepository getGitRepository() {
        Optional<UserPassword> credentials = getCredentials();
        return new GitRepository(
                getType(),
                getName(),
                getRemote(),
                credentials.map(UserPassword::getUser).orElse(""),
                credentials.map(UserPassword::getPassword).orElse("")
        );
    }

}

File: ontrack-extension-git/src/main/java/net/nemerosa/ontrack/extension/git/model/GitConfiguration.kt
package net.nemerosa.ontrack.extension.git.model

import net.nemerosa.ontrack.extension.issues.model.ConfiguredIssueService
import net.nemerosa.ontrack.git.GitRepositoryAuthenticator

/**
 * Definition of a Git configuration.
 */
interface GitConfiguration {

    /**
     * Type
     */
    val type: String

    /**
     * Name in the type
     */
    val name: String

    /**
     * Remote URL-ish
     */
    val remote: String

    /**
     * Authentication mode
     */
    val authenticator: GitRepositoryAuthenticator?

    /**
     * Link to a commit, using {commit} as placeholder
     */
    val commitLink: String

    /**
     * Link to a file at a given commit, using {commit} and {path} as placeholders
     */
    val fileAtCommitLink: String

    /**
     * Indexation interval
     */
    val indexationInterval: Int

    /**
     * Gets the associated issue service configuration (if any)
     */
    val configuredIssueService: ConfiguredIssueService?

}
